taichi/util/offline_cache.h

```diff
@@ -0,0 +1,248 @@
+#pragma once
+
+#include <ctime>
+#include <cstdint>
+#include <queue>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+
+#include "taichi/common/core.h"
+#include "taichi/common/cleanup.h"
+#include "taichi/util/io.h"
+#include "taichi/util/lock.h"
+
+namespace taichi {
+namespace lang {
+namespace offline_cache {
+
+using Version = std::uint16_t[3];  // {MAJOR, MINOR, PATCH}
+
+enum CleanCacheFlags {
+  NotClean = 0b000,
+  CleanOldVersion = 0b001,
+  CleanOldUsed = 0b010,
+  CleanOldCreated = 0b100
+};
+
+enum CleanCachePolicy {
+  Never = NotClean,
+  OnlyOldVersion = CleanOldVersion,
+  LRU = CleanOldVersion | CleanOldUsed,
+  FIFO = CleanOldVersion | CleanOldCreated
+};
+
+inline CleanCachePolicy string_to_clean_cache_policy(const std::string &str) {
+  if (str == "never") {
+    return Never;
+  } else if (str == "version") {
+    return OnlyOldVersion;
+  } else if (str == "lru") {
+    return LRU;
+  } else if (str == "fifo") {
+    return FIFO;
+  }
+  return Never;
+}
+
+template <typename BackendSpecificMetadata>
+struct KernelMetadata {
```

Comment:
`KernelMetadata` and `Metadata` doesn't seem to be used anywhere. Do you plan to rewrite `KernelCacheData` with this? BTW why do you choose to add the `extend_data` field instead of making `KernelMetadata` a virtual class?

Reply:
Thanks for your suggestion. `KernelMetadata` and `Metadata` will be used to implement cache cleaning on other backends.
I removed them in https://github.com/taichi-dev/taichi/pull/5934/commits/a2244c9fddaa042fd0faa67b1050551311267590 temporarily, please review again.